src/Dhall/Map.hs

```diff
@@ -0,0 +1,258 @@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
+    deriving (Data, Eq, Foldable, Functor, Show, Traversable)
+
+instance Ord k => Data.Semigroup.Semigroup (Map k v) where
+    (<>) = union
+
+instance Ord k => Monoid (Map k v) where
+    mempty = Map Data.Map.empty []
+
+#if !(MIN_VERSION_base(4,11,0))
+    mappend = (<>)
+#endif
+
+-- | Create a `Map` from a single key-value pair
+singleton :: k -> v -> Map k v
+singleton k v = Map m ks
+  where
+    m = Data.Map.singleton k v
+
+    ks = pure k
+
+{-| Create a `Map` from a list of key-value pairs
+
+> fromList empty = mempty
+>
+> fromList (x <|> y) = fromList x <> fromList y
+-}
+fromList :: Ord k => [(k, v)] -> Map k v
+fromList kvs = Map m ks
+  where
+    m = Data.Map.fromList kvs
+
+    ks = nub (map fst kvs)
+
+nub :: Ord k => [k] -> [k]
+nub = go Data.Set.empty
+  where
+    go _      []  = []
+    go set (k:ks)
+        | Data.Set.member k set =     go                    set  ks
+        | otherwise             = k : go (Data.Set.insert k set) ks
+
+{-| Sort the keys of a `Map`, forgetting the original ordering
+
+> sort (sort x) = sort x
+-}
+sort :: Ord k => Map k v -> Map k v
+sort (Map m _) = Map m ks
+  where
+    ks = Data.Map.keys m
+
+{-| Check if the keys of a `Map` are already sorted
+
+> isSorted (sort m) = True
+-}
+isSorted :: Eq k => Map k v -> Bool
+isSorted (Map m k) = Data.Map.keys m == k
+
+{-| Insert a key-value pair into a `Map`, overriding any previous value stored
+    underneath the same key, if present
+
+> insert = insertWith (\v _ -> v)
+-}
+insert :: Ord k => k -> v -> Map k v -> Map k v
+insert k v (Map m ks) = Map m' ks'
+  where
+    m' = Data.Map.insert k v m
+
+    ks' | elem k ks = ks
+        | otherwise = k : ks
+
+{-| Insert a key-value pair into a `Map`, using the supplied function to combine
+    the new value with any old value underneath the same key, if present
+-}
+insertWith :: Ord k => (v -> v -> v) -> k -> v -> Map k v -> Map k v
+insertWith f k v (Map m ks) = Map m' ks'
+  where
+    m' = Data.Map.insertWith f k v m
+
+    ks' | elem k ks = ks
+        | otherwise = k : ks
+
+-- | Delete a key from a `Map` if present, otherwise return the original `Map`
+delete :: Ord k => k -> Map k v -> Map k v
+delete k (Map m ks) = Map m' ks'
+  where
+    m' = Data.Map.delete k m
+
+    ks' = filter (k /=) ks
+
+{-| Retrieve a key from a `Map`
+
+> lookup k mempty = empty
+>
+> lookup k (x <> y) = lookup k y <|> lookup k x
+-}
+lookup :: Ord k => k -> Map k v -> Maybe v
+lookup k (Map m _) = Data.Map.lookup k m
+
+{-| Check if a key belongs to a `Map`
+
+> member k mempty = False
+>
+> member k (x <> y) = member k x || member k y
+-}
+member :: Ord k => k -> Map k v -> Bool
+member k (Map m _) = Data.Map.member k m
+
+{-| Combine two `Map`s, preferring keys from the first `Map`
+
+> union = unionWith (\v _ -> v)
+-}
+union :: Ord k => Map k v -> Map k v -> Map k v
+union (Map mL ksL) (Map mR ksR) = Map m ks
```

Comment:
Why not define in terms of `unionWith`?

Reply:
@quasicomputational: This was just in case `Data.Map.union` was more efficient implementation than `Data.Map.unionWith const`